math/src/polynom/multi_linear.rs

```diff
@@ -0,0 +1,240 @@
+    num_variables: usize,
+    evaluations: Vec<E>,
+}
+
+impl<E: FieldElement> MultiLinearPoly<E> {
+    /// Constructs a [`MultiLinearPoly`] from its evaluations over the boolean hyper-cube {0 , 1}^ν.
+    pub fn from_evaluations(evaluations: Vec<E>) -> Result<Self, MultiLinearPolyError> {
+        if !evaluations.len().is_power_of_two() {
+            return Err(MultiLinearPolyError::EvaluationsNotPowerOfTwo);
+        }
+        Ok(Self {
+            num_variables: (evaluations.len().ilog2()) as usize,
+            evaluations,
+        })
+    }
+
+    /// Returns the number of variables of the multi-linear polynomial.
+    pub fn num_variables(&self) -> usize {
+        self.num_variables
+    }
+
+    /// Returns the evaluations over the boolean hyper-cube.
+    pub fn evaluations(&self) -> &[E] {
+        &self.evaluations
+    }
+
+    /// Returns the number of evaluations. This is equal to the size of the boolean hyper-cube.
+    pub fn num_evaluations(&self) -> usize {
+        self.evaluations.len()
+    }
+
+    /// Evaluate the multi-linear at some query (r_0, ..., r_{ν - 1}) ∈ 𝔽^ν.
+    ///
+    /// It first computes the evaluations of the Lagrange basis polynomials over the interpolating
+    /// set {0 , 1}^ν at (r_0, ..., r_{ν - 1}) i.e., the Lagrange kernel at (r_0, ..., r_{ν - 1}).
+    /// The evaluation then is the inner product, indexed by {0 , 1}^ν, of the vector of
+    /// evaluations times the Lagrange kernel.
+    pub fn evaluate(&self, query: &[E]) -> E {
+        let tensored_query = compute_lagrange_basis_evals_at(query);
+        inner_product(self.evaluations.iter().copied(), tensored_query.iter().copied())
+    }
+
+    /// Similar to [`Self::evaluate`], except that the query was already turned into the Lagrange
+    /// kernel (i.e. the [`lagrange_ker::EqFunction`] evaluated at every point in the set
+    /// `{0 , 1}^ν`).
+    ///
+    /// This is more efficient than [`Self::evaluate`] when multiple different [`MultiLinearPoly`]
+    /// need to be evaluated at the same query point.
+    pub fn evaluate_with_lagrange_kernel(&self, lagrange_kernel: &[E]) -> E {
+        inner_product(self.evaluations.iter().copied(), lagrange_kernel.iter().copied())
+    }
+
+    /// Computes f(r_0, y_1, ..., y_{ν - 1}) using the linear interpolation formula
+    /// (1 - r_0) * f(0, y_1, ..., y_{ν - 1}) + r_0 * f(1, y_1, ..., y_{ν - 1}) and assigns
+    /// the resulting multi-linear, defined over a domain of half the size, to `self`.
+    pub fn bind_least_significant_variable(&mut self, round_challenge: E) {
+        let mut result = vec![E::ZERO; 1 << (self.num_variables() - 1)];
+        for (i, res) in result.iter_mut().enumerate() {
+            *res = self.evaluations[i << 1]
+                + round_challenge * (self.evaluations[(i << 1) + 1] - self.evaluations[i << 1]);
+        }
+        *self = Self::from_evaluations(result)
+            .expect("should not fail given that it is a multi-linear");
+    }
```

Comment:
I find this function a little confusing. I think what we are doing is reducing the number of variables by 1 and as the result, halving the number of evaluations - right?

If so, we should probably just return a new `MultiLinearPoly` from here. For example, the signature could be:
```rust
pub fn bind_least_significant_variable(self, variable_value: E) -> Self
```

Another thing, it seems to me that we don't actually need to allocate the `result` vector. Instead, we may be able to update the `self.evaluations` vector in place and then just truncate its length. Or am I missing something?

Reply:
Are the two suggestions compatible? I went with the second one and removed the need for `result`